Source/Core/RequestScheduler.js

```diff
@@ -129,41 +256,93 @@ define([
+        if (!defined(request.requestFunction)) {
+            throw new DeveloperError('request.requestFunction is required.');
         }
         //>>includeEnd('debug');
 
-        if (activeRequests >= RequestScheduler.maximumRequests) {
-            return undefined;
+        ++stats.numberOfRequestsThisFrame;
+
+        if (!RequestScheduler.throttle) {
+            return request.requestFunction(request.url, request.parameters);
         }
 
-        var server = RequestScheduler.getServer(url);
-        var activeRequestsForServer = defaultValue(activeRequestsByServer[server], 0);
-        if (activeRequestsForServer >= RequestScheduler.maximumRequestsPerServer) {
-            return undefined;
+        var server = RequestScheduler.getServer(request.url);
+        request.server = server;
+        activeRequestsByServer[server] = defaultValue(activeRequestsByServer[server], 0);
+
+        if (!RequestScheduler.hasAvailableRequests(request.url)) {
+            if (!request.defer) {
+                // No available slots to make the request, return undefined
+                handleLeftoverRequest(request);
+                return undefined;
+            } else {
+                // If no slots are available, the request is deferred until a slot opens up.
+                // Return a promise even if the request can't be completed immediately.
+                return deferRequest(request);
+            }
         }
 
-        ++activeRequests;
-        activeRequestsByServer[server] = activeRequestsForServer + 1;
+        if (RequestScheduler.prioritize && defined(request.requestType) && !request.defer) {
+            var budget = getBudget(request);
+            if (budget.used >= budget.total) {
+                // Request does not fit in the budget, return undefined
+                handleLeftoverRequest(request);
+                return undefined;
+            }
+            ++budget.used;
+        }
 
-        return when(requestFunction(url), function(result) {
-            --activeRequests;
-            --activeRequestsByServer[server];
-            return result;
-        }).otherwise(function(error) {
-            --activeRequests;
-            --activeRequestsByServer[server];
-            return when.reject(error);
-        });
+        return startRequest(request);
+    };
+
+    /**
+     * A function that will make a request when an open slot is available. Always returns
+     * a promise, which is suitable for data sources and utility functions.
+     *
+     * @param {String} url The URL to request.
+     * @param {RequestScheduler~RequestFunction} requestFunction The actual function that
+     *        makes the request.
+     * @param {Object} [parameters] Extra parameters to send with the request.
+     *
+     * @returns {Promise.<Object>} A Promise for the requested data.
+     */
+    RequestScheduler.request = function(url, requestFunction, parameters) {
+        return RequestScheduler.throttleRequest(new Request({
```

Comment:
Perhaps this should have a `type` of "OTHER", "USER", "CUSTOM", or something like that, which could be useful for setting priorities and avoiding starvation.